src/cargo/ops/cargo_compile/mod.rs

```diff
@@ -602,6 +623,105 @@ where `<compatible-ver>` is the latest version supporting rustc {rustc_version}"
     Ok(bcx)
 }
 
+// Checks if a target path exists and is a source file, not a directory
+fn validate_target_path_as_source_file(
+    gctx: &GlobalContext,
+    target_path: &std::path::Path,
+    target_name: &str,
+    target_kind: &TargetKind,
+    unit_manifest_path: &std::path::Path,
+    error_count: &mut usize,
+) -> CargoResult<()> {
+    if !target_path.exists() {
+        *error_count += 1;
+
+        let err_msg = format!(
+            "can't find {} `{}` at path `{}`",
+            target_kind.description(),
+            target_name,
+            target_path.display()
+        );
+
+        let group = Group::with_title(Level::ERROR.primary_title(err_msg)).element(Origin::path(
+            unit_manifest_path.to_str().unwrap_or_default(),
+        ));
+
+        gctx.shell().print_report(&[group], true)?;
+    } else if target_path.is_dir() {
+        *error_count += 1;
+
+        // suggest setting the path to a likely entrypoint
+        let main_rs = target_path.join("main.rs");
+        let lib_rs = target_path.join("lib.rs");
+
+        let suggested_files_opt = match target_kind {
+            TargetKind::Lib(_) => {
+                if lib_rs.exists() {
+                    Some(format!("`{}`", lib_rs.display()))
+                } else {
+                    None
+                }
+            }
+            TargetKind::Bin => {
+                if main_rs.exists() {
+                    Some(format!("`{}`", main_rs.display()))
+                } else {
+                    None
+                }
+            }
+            TargetKind::Test => {
+                if main_rs.exists() {
+                    Some(format!("`{}`", main_rs.display()))
+                } else {
+                    None
+                }
+            }
+            TargetKind::ExampleBin => {
+                if main_rs.exists() {
+                    Some(format!("`{}`", main_rs.display()))
+                } else {
+                    None
+                }
+            }
+            TargetKind::Bench => {
+                if main_rs.exists() {
+                    Some(format!("`{}`", main_rs.display()))
+                } else {
+                    None
+                }
+            }
+            TargetKind::ExampleLib(_) => {
+                if lib_rs.exists() {
+                    Some(format!("`{}`", lib_rs.display()))
+                } else {
+                    None
+                }
+            }
+            _ => None,
```

Comment:
Can we explicitly enumerate all `TargetKind`s so the compiler can remind us to re-evaluate this if a new variant is added?